Compiler error with escaped double-quote

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

**V version:** 2cb711e
**OS:** Win10

**What did you do?**
Tried to compile this file:
```
fn main() {
	s := "escaped double-quote \""
	println(s)
}
```

**What did you expect to see?**
That it compiles.
 
**What did you see instead?**
I get following error:
```
D:\src\v>v error.v
builder error: Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl -w /we4013 /volatile:ms /Fo"C:\temp\v\error.tmp.c.obj" /MDd "C:\temp\v\error.tmp.c" -I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt" -I "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.26.28801\include" -I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um" -I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared" kernel32.lib user32.lib advapi32.lib dbghelp.lib /link /NOLOGO /OUT:"D:\src\v\error.exe" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\X64" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\X64" /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.26.28801\lib\X64" /DEBUG:FULL

error.tmp.c
C:\temp\v\error.tmp.c(8010): error C2146: syntax error: missing ')' before identifier 'bar'
C:\temp\v\error.tmp.c(8010): error C2017: illegal escape sequence
C:\temp\v\error.tmp.c(8010): error C2059: syntax error: ')'
```

--- a/vlib/builtin/string_test.v
+++ b/vlib/builtin/string_test.v
@@ -619,7 +619,7 @@ fn test_for_loop_two() {
 
 fn test_quote() {
 	a := `'`
-	println("testing double quotes")
+	println("testing double quotes \"")
 	b := "hi"
 	assert b == 'hi'
 	assert a.str() == '\''

--- a/vlib/v/gen/str.v
+++ b/vlib/v/gen/str.v
@@ -5,6 +5,33 @@ module gen
 import v.ast
 import v.table
 
+fn escape_string(str string, raw, quotes, newlines, percentage, double_escape bool) string {
+	mut escaped_val := str
+	if quotes {
+		escaped_val = escaped_val.replace_each(['"', '\\"'])
+	}
+	if raw {
+		return escaped_val
+	}
+	if double_escape {
+		escaped_val = escaped_val.replace_each(['\\', '\\\\'])
+	}
+	if !percentage && !double_escape {
+		escaped_val = escaped_val.replace_each(['\\\\"', '\\042'])
+	}
+	if newlines {
+		escaped_val = escaped_val.replace_each(['\r\n', '\\n', '\n', '\\n'])
+	}
+	if percentage {
+		escaped_val = escaped_val.replace_each(['%', '%%'])
+	}
+	return escaped_val
+}
+
+fn escape_all(str string) string {
+	return escape_string(str, true, true, true, true, true)
+}
+
 fn (mut g Gen) write_str_fn_definitions() {
 	// _STR function can't be defined in vlib
 	g.writeln("
@@ -118,11 +145,11 @@ string _STR_TMP(const char *fmt, ...) {
 
 fn (mut g Gen) string_literal(node ast.StringLiteral) {
 	if node.is_raw {
-		escaped_val := node.val.replace_each(['"', '\\"', '\\', '\\\\'])
+		escaped_val := escape_string(node.val, true, true, false, false, true)
 		g.write('tos_lit("$escaped_val")')
 		return
 	}
-	escaped_val := node.val.replace_each(['"', '\\"', '\r\n', '\\n', '\n', '\\n'])
+	escaped_val := escape_string(node.val, false, true, true, false, false)
 	if g.is_c_call || node.language == .c {
 		// In C calls we have to generate C strings
 		// `C.printf("hi")` => `printf("hi");`
@@ -149,7 +176,7 @@ fn (mut g Gen) string_inter_literal_sb_optimized(call_expr ast.CallExpr) {
 	is_nl := call_expr.name == 'writeln'
 	// println('optimize sb $call_expr.name')
 	for i, val in node.vals {
-		escaped_val := val.replace_each(['"', '\\"', '\r\n', '\\n', '\n', '\\n', '%', '%%'])
+		escaped_val := escape_string(val, false, true, true, true, false)
 		// if val == '' {
 		// break
 		// continue
@@ -220,7 +247,7 @@ fn (mut g Gen) string_inter_literal(node ast.StringInterLiteral) {
 	// Build the string with %
 	mut end_string := false
 	for i, val in node.vals {
-		escaped_val := val.replace_each(['"', '\\"', '\r\n', '\\n', '\n', '\\n', '%', '%%'])
+		escaped_val := escape_string(val, false, true, true, false, false)
 		if i >= node.exprs.len {
 			if escaped_val.len > 0 {
 				end_string = true

--- a/vlib/v/gen/str.v
+++ b/vlib/v/gen/str.v
@@ -5,32 +5,91 @@ module gen
 import v.ast
 import v.table
 
-fn escape_string(str string, raw, quotes, newlines, percentage, double_escape bool) string {
-	mut escaped_val := str
-	if quotes {
-		escaped_val = escaped_val.replace_each(['"', '\\"'])
+fn smart_quote(str string, raw bool) string {
+	len := str.len
+	if len == 0 {
+		return str
 	}
-	if raw {
-		return escaped_val
+	mut result := ''
+	mut pos := -1
+	mut last := ''
+	mut next := ''
+	mut skip_next := false
+	for {
+		pos = pos + 1
+		if skip_next {
+			skip_next = false
+			pos = pos + 1
+		}
+		if pos >= len {
+			break
+		}
+		if pos + 1 < len {
+			next = str.str[pos + 1].str()
+		}
+		mut current := str
+		mut toadd := str
+		if len > 1 {
+			current = str.str[pos].str()
+			toadd = current
+		}
+		// double quote
+		if current == '"' {
+			toadd = '\\"'
+			current = ''
+		}
+		if raw && current == '\\' {
+			toadd = '\\\\'
+		}
+		// keep newlines in string
+		if current == '\n' {
+			toadd = '\\n'
+			current = ''
+		}
+		if current == '\r' && next == '\n' {
+			toadd = '\r\n'
+			current = ''
+			skip_next = true
+		}
+		// backslash
+		if !raw && current == '\\' {
+			// escaped backslash - keep as is
+			if next == '\\' {
+				toadd = '\\\\'
+				skip_next = true
+			}
+			// keep raw escape squence
+			else {
+				if next != '' {
+					if raw {
+						toadd = '\\\\' + next
+						skip_next = true
+					}
+					// escape it
+					else {
+						toadd = '\\' + next
+						skip_next = true
+					}
+				}
+			}
+		}
+		// Dolar sign
+		if !raw && current == '$' {
+			if last == '\\' {
+				toadd = '\\$'
+			}
+		}
+		// Windows style new line \r\n
+		if !raw && current == '\r' {
+			if next == '\n' {
+				skip_next = true
+				toadd = '\\n'
+			}
+		}
+		result = result + toadd
+		last = current
 	}
-	if double_escape {
-		escaped_val = escaped_val.replace_each(['\\', '\\\\'])
-	}
-	if quotes && !percentage && !double_escape {
-		escaped_val = escaped_val.replace_each(['"""', '"\\042"'])
-		escaped_val = escaped_val.replace_each(['\\\\"', '\\042'])
-	}
-	if newlines {
-		escaped_val = escaped_val.replace_each(['\r\n', '\\n', '\n', '\\n'])
-	}
-	if percentage {
-		escaped_val = escaped_val.replace_each(['%', '%%'])
-	}
-	return escaped_val
-}
-
-fn escape_all(str string) string {
-	return escape_string(str, false, true, true, true, true)
+	return result
 }
 
 fn (mut g Gen) write_str_fn_definitions() {
@@ -146,11 +205,11 @@ string _STR_TMP(const char *fmt, ...) {
 
 fn (mut g Gen) string_literal(node ast.StringLiteral) {
 	if node.is_raw {
-		escaped_val := escape_string(node.val, true, true, false, false, true)
+		escaped_val := smart_quote(node.val, true)
 		g.write('tos_lit("$escaped_val")')
 		return
 	}
-	escaped_val := escape_string(node.val, false, true, true, false, false)
+	escaped_val := smart_quote(node.val, false)
 	if g.is_c_call || node.language == .c {
 		// In C calls we have to generate C strings
 		// `C.printf("hi")` => `printf("hi");`
@@ -177,7 +236,7 @@ fn (mut g Gen) string_inter_literal_sb_optimized(call_expr ast.CallExpr) {
 	is_nl := call_expr.name == 'writeln'
 	// println('optimize sb $call_expr.name')
 	for i, val in node.vals {
-		escaped_val := escape_string(val, false, true, true, true, false)
+		escaped_val := smart_quote(val, false)
 		// if val == '' {
 		// break
 		// continue
@@ -248,7 +307,8 @@ fn (mut g Gen) string_inter_literal(node ast.StringInterLiteral) {
 	// Build the string with %
 	mut end_string := false
 	for i, val in node.vals {
-		escaped_val := escape_string(val, false, true, true, false, false)
+		mut escaped_val := val.replace_each(['%', '%%'])
+		escaped_val = smart_quote(escaped_val, false)
 		if i >= node.exprs.len {
 			if escaped_val.len > 0 {
 				end_string = true

--- a/vlib/v/gen/str.v
+++ b/vlib/v/gen/str.v
@@ -16,7 +16,8 @@ fn escape_string(str string, raw, quotes, newlines, percentage, double_escape bo
 	if double_escape {
 		escaped_val = escaped_val.replace_each(['\\', '\\\\'])
 	}
-	if !percentage && !double_escape {
+	if quotes && !percentage && !double_escape {
+		escaped_val = escaped_val.replace_each(['"""', '"\\042"'])
 		escaped_val = escaped_val.replace_each(['\\\\"', '\\042'])
 	}
 	if newlines {
@@ -29,7 +30,7 @@ fn escape_string(str string, raw, quotes, newlines, percentage, double_escape bo
 }
 
 fn escape_all(str string) string {
-	return escape_string(str, true, true, true, true, true)
+	return escape_string(str, false, true, true, true, true)
 }
 
 fn (mut g Gen) write_str_fn_definitions() {

--- a/vlib/v/gen/str.v
+++ b/vlib/v/gen/str.v
@@ -25,12 +25,16 @@ fn smart_quote(str string, raw bool) string {
 			break
 		}
 		if pos + 1 < len {
-			next = str.str[pos + 1].str()
+			unsafe {
+				next = str.str[pos + 1].str()
+			}
 		}
 		mut current := str
 		mut toadd := str
 		if len > 1 {
-			current = str.str[pos].str()
+			unsafe {
+				current = str.str[pos].str()
+			}
 			toadd = current
 		}
 		// double quote

